scripts/release.sh

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env bash
+PREV_BRANCH=$(git symbolic-ref HEAD --short)
+IS_DIRTY=$(git diff --ignore-submodules HEAD)
+
+VERSION_TYPE=""
+VERSION_TYPE_MESSAGE="If specifying a release type argument, please provide 'major', 'minor', or 'patch'"
+
+NEW_VERSION=""
+BRANCH_NAME="release/0.0.0"
+CURR_VERSION=$(npm run env | grep npm_package_version | cut -f 2 -d '=')
+MAJOR_VERSION=$(semver $CURR_VERSION -i major)
+MINOR_VERSION=$(semver $CURR_VERSION -i minor)
+PATCH_VERSION=$(semver $CURR_VERSION -i patch)
+
+function stash_checkout_version() {
+  # Stash changes if branch is dirty
+  if [ -n "$IS_DIRTY" ]; then
+    echo "Current branch is dirty. Stashing changes..."
+    git stash --quiet
+  fi
+
+  # Cut release branch from master
+  echo "Checking out master and pulling latest..."
+  git checkout master --quiet
+  git pull origin master --quiet
+
+  git checkout -b "$BRANCH_NAME" --quiet
+  echo "Bumping version to $NEW_VERSION"
+  npm version $VERSION_TYPE
+
+  echo "Publishing..."
+  npm publish
+
+  echo "Pushing version bump and tags..."
+  git push origin "$BRANCH_NAME" --quiet
+  git push origin "$BRANCH_NAME" --tags --quiet
+
+  # Return to branch
+  git checkout $PREV_BRANCH --quiet
+
+  # Reapply stashed changes if branch was dirty
+  if [ -n "$IS_DIRTY" ]; then
+    git stash pop --quiet
+  fi
+}
+
+function get_version() {
+  case $1 in
+    "major" | "1")
+      VERSION_TYPE="major"
+      NEW_VERSION=$MAJOR_VERSION
+      ;;
+    "minor" | "2")
+      VERSION_TYPE="minor"
+      NEW_VERSION=$MINOR_VERSION
+      ;;
+    "patch" | "3")
+      VERSION_TYPE="patch"
+      NEW_VERSION=$PATCH_VERSION
+      ;;
+    *)
+      echo $VERSION_TYPE_MESSAGE
+      exit
+  esac
+  BRANCH_NAME="release/$NEW_VERSION"
+}
+
+echo "Current package version is $CURR_VERSION"
+if [ -n "$1" ]; then
+  get_version $1
+else
+  echo "Please select release type:"
+  echo "   1) major ($MAJOR_VERSION)"
+  echo "   2) minor ($MINOR_VERSION)"
+  echo "   3) patch ($PATCH_VERSION)"
+  read RESPONSE
+  get_version $RESPONSE
+fi
+echo "Creating a $VERSION_TYPE ($NEW_VERSION) release"
+
+stash_checkout_version
```

Comment:
Great script man! I'm good with all this. 

Only thought I had was that where the pull from master occurs is after the current version is read. If master isn't up to date, or the branch current checked out has a stale version, the version created could be incorrect. Totally fine to leave it as is.

Reply:
Ahh good point. Will find a fix for that tomorrow.